File: javascript/src/hangman.js
const input = require('sync-input');

let langs = ["python", "java", "swift", "javascript"];
const lowercase = new RegExp('[a-z]');
let continuePlaying = true;
let attemptsLeft;
let word;
let filled;
let revealed;
let lettersGuessed;
let wins = 0;
let losses = 0;

let setCharAt = (str, index, chr) => str.substring(0, index) + chr + str.substring(1 + index, str.length);

let displayMenu = () => {
    while (continuePlaying) {
        let choice = input('Type "play" to play the game, "results" to show the scoreboard, and "exit" to quit:');
        switch (choice) {
            case "play":
                reset();
                playGame();
                break;
            case "results":
                displayResult();
                break;
            case "exit":
                continuePlaying = false;
                break;
        }
    }
}
let reset = () => {
    attemptsLeft = 8;
    word = langs[Math.floor((Math.random() * langs.length))];
    filled = '-'.repeat(word.length);
    revealed = '';
    lettersGuessed = '';
    console.log();
}

let playGame = () => {
    console.log(filled);
    while (attemptsLeft > 0) {
        let letter = input("Input a letter: ");
        if (isValidInput(letter)) {
            lettersGuessed = lettersGuessed + letter;
            handleGuess(letter);
        }
        console.log();
        if (attemptsLeft > 0) console.log(filled);
    }
    if (revealed != null && revealed === word) {
        wins++;
        console.log(`You guessed the word ${word}!`);
        console.log("You survived!");
    } else {
        losses++;
        console.log("You lost!");
    }
}

let displayResult = () => {
    console.log(`You won: ${wins} times.`);
    console.log(`You lost: ${losses} times.`);
}
let isValidInput = (guess) => {
    let isValid = true;
    if (guess.length !== 1) {
        isValid = false;
        console.log("Please, input a single letter.");
    } else if (!lowercase.test(guess)) {
        isValid = false;
        console.log("Please, enter a lowercase letter from the English alphabet.");
    } else if (lettersGuessed.length > 0 && lettersGuessed.indexOf(guess) > -1) {
        isValid = false;
        console.log("You've already guessed this letter.");
    }
    return isValid;
}

let updateFilled = (guess) => {
    for (let i = 0; i < word.length; i++) {
        if (filled[i] !== '-') continue;
        if (word[i] === guess) {
            filled = setCharAt(filled, i, guess);
        }
    }
}

let handleGuess = (guess) => {
    if (word.indexOf(guess) < 0) {
        console.log("That letter doesn't appear in the word.");
        attemptsLeft--;
    } else if (revealed.length > 0 && revealed.indexOf(guess) > -1) {
        console.log("No improvements.");
        attemptsLeft--;
    } else {
        updateFilled(guess);
        revealed = filled.replaceAll('-', '');
        if (revealed === word || filled.indexOf('-') < 0) attemptsLeft = 0;
    }
}

console.log('H A N G M A N');
displayMenu();

File: python/projects/hangman.py
import os
import random
import re


def display_menu():
    print('H A N G M A N\n')
    params = {}
    results = {"wins": 0, "losses": 0}
    while True:
        choice = input('Type "play" to play the game, "results" to show the scoreboard, '
                       'and "exit" to quit:\n')
        match choice:
            case "play":
                params = reset(random.choice(words))
                print(''.join(params['filled']))
                if play_game(params):
                    print(f'You guessed the word {params["word"]}!\nYou survived!\n')
                    results['wins'] += 1
                else:
                    print('You lost!\n')
                    results['losses'] += 1
            case "results":
                display_results(results)
            case "exit":
                params['exit'] = True
                break
    return params


def reset(word):
    return {
        'attempts_left': 8,
        'filled': ['-' for x in word],
        'word': word,
        'guesses': [],
        'exit': False
    }


def play_game(params):
    while params['attempts_left'] > 0:
        letter = input('Input a letter:')
        inc_attempt = handle_guess(params, letter)
        params['attempts_left'] -= inc_attempt
        if ''.join(params['filled']) == params['word']:
            return True
        if params['attempts_left'] > 0:
            print(''.join(params['filled']))
    return False


def update_filled(params, guess):
    for i in range(0, len(params['word'])):
        if params['filled'][i] != '-':
            continue
        if params['word'][i] == guess:
            params['filled'][i] = guess


def is_valid_letter(guessed):
    if len(guessed) != 1:
        print('Please, input a single letter.\n')
        return False
    if not re.match('^[a-z]*$', guessed):
        print('Please, enter a lowercase letter from the English alphabet.\n')
        return False
    return True


def handle_guess(params, guess):
    if not is_valid_letter(guess):
        return 0
    if guess in params['guesses']:
        print("You've already guessed this letter.\n")
    elif guess in params['word']:
        update_filled(params, guess)
        print()
    else:
        params['guesses'].append(guess)
        print("That letter doesn't appear in the word.\n")
        return 1
    params['guesses'].append(guess)
    return 0


def display_results(results):
    print(f'You won: {results["wins"]} times\nYou lost: {results["losses"]} times')


words = ['python', 'java', 'swift', 'javascript']
game_params = display_menu()

